src/main/java/Links.java

```diff
@@ -1,5 +1,139 @@
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 interface Links extends Iterable<URL> {
+    Iterator<URL> iterator();
+    String toString();
+
     class HTML implements Links {
+        private String inputURL;
+        private ArrayList<URL> links;
+        private HashMap<String, ArrayList<URL>> deadLinks;
+        public String errorStatus;
+
+        private boolean isValidLink(String url) {
+            String urlRegex = "^(http|https)://[-a-zA-Z0-9+&@#/%?=~_|,!:.;]*[-a-zA-Z0-9+@#/%=&_|]";
+            Pattern pattern = Pattern.compile(urlRegex);
+            Matcher m = pattern.matcher(url);
+            if (m.matches()) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        private ArrayList<URL> retrieveLinks() {
+            ArrayList<URL> tempLinks = new ArrayList<>();
+            try {
+                Document document = Jsoup.connect(this.inputURL.toString()).get();
+                Elements documentLinks = document.select("a[href]");
+                for (Element link : documentLinks) {
+                    URL tempLink = new URL(link.attr("abs:href").toString());
+                    if (isValidLink(tempLink.toString())) {
+                        tempLinks.add(tempLink);
+                    }
+                }
+            } catch (MalformedURLException e) {
+                errorStatus = "One of the retrieved links is malformed";
+            } catch (IOException e) {
+                errorStatus = "IOException while connecting";
+            } catch (IllegalArgumentException e) {
+                errorStatus = "Error while reading parent URL.";
+            }
+            return tempLinks;
+        }
+
+        private int getStatus(URL url) {
+            try {
+                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+                connection.connect();
+                int statusCode = connection.getResponseCode();
+                return statusCode;
+            } catch (IOException e) {
+                return 0;
+            }
+        }
+
+        private void getDeadLinks() {
```

Comment:
Even though, it is currently used in a `void` way, I'd think of further extensibility and translate it into a public method returning the hash map, so that you may properly consume processed result from outside not only in a string form (@ekondrashev, please share your thoughts here)